File: src/main/java/algostrategies/dp08/Numb3rs.java
package algostrategies.dp08;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Numb3rs {

    private static BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));

    private static double[][] cache;
    private static int[] degree;
    private static int prison;
    private static int townSize;
    private static int[][] town;

    public static void main(String[] args) throws IOException {
        int cases = Integer.parseInt(stdin.readLine());
        while (cases-- > 0) {
            String[] targets = stdin.readLine().split(" ");
            int day = Integer.parseInt(targets[1]);

            townSize = Integer.parseInt(targets[0]);
            prison = Integer.parseInt(targets[2]);
            degree = new int[townSize];

            cache = new double[townSize + 1][day + 1];
            for (int t = 0; t < townSize + 1; t++) {
                for (int d = 0; d < day + 1; d++) {
                    cache[t][d] = -1.0;
                }
            }

            town = new int[townSize][townSize];
            for (int y = 0; y < townSize; y++) {
                String[] cells = stdin.readLine().split(" ");
                int deg = 0;

                for (int x = 0; x < townSize; x++) {
                    town[y][x] = Integer.parseInt(cells[x]);
                    deg += town[y][x];
                }

                degree[y] = deg;
            }

            int problem = Integer.parseInt(stdin.readLine());
            String[] problems = stdin.readLine().split(" ");
            StringBuffer result = new StringBuffer();

            for (int p = 0; p < problem; p++) {
                double r = solve(Integer.parseInt(problems[p]), day);
                result
                    .append(String.format(r + "", "%.8f"))
                    .append(" ");
            }

            System.out.print(result);
        }
    }

    /**
     * Java Implementation of Numb3rs.scala
     */
    private static double solve(int here, int day) {
        if (day == 0) {
            return here == prison ? 1.0 : 0.0;
        } else if (cache[here][day] > -0.5) {
            return cache[here][day];
        }

        double result = 0;

        for (int there = 0; there < townSize; there++) {
            if (town[there][here] == 1) {
                result += solve(there, day - 1) / degree[there];
            }
        }

        cache[here][day] = result;
        return result;
    }
}

File: src/main/scala/algostrategiesscala/dp08/Numb3rs.scala
package algostrategiesscala.dp08

import java.io.{BufferedReader, InputStreamReader}

/**
  * Created by honey.and.sw on 2017. 9. 24.
  */
object Numb3rs {

  val stdin = new BufferedReader(new InputStreamReader(System.in))

  def main(args: Array[String]): Unit = {
    val cases = Integer.parseInt(stdin.readLine())

    for (_ <- 0 until cases) {
      val target = stdin.readLine()
      val targets = target.split(" ")

      val townSize = targets(0).toInt
      val day = targets(1).toInt
      val prison = targets(2).toInt

      val town = Array.fill(townSize)(Array.fill(townSize)(0))
      for {
        y <- 0 until townSize
      } {
        val cells: Array[Int] = stdin.readLine().split(" ").map(_.toInt)
        town(y) = cells
      }

      val problemSize = Integer.parseInt(stdin.readLine())
      val result: Array[String] = stdin.readLine().split(" ")
          .map(_.toInt)
          .map { village =>
            Problem(town, prison, village, day).solve.formatted("%.8f")
          }

      println(result.mkString(" "))
    }
  }

  case class Problem(town: Array[Array[Int]], prison: Int, village: Int, day: Int) {

    val neighbors: Array[Int] = Array.fill(town.length)(0)
    val cache: Array[Array[Double]] = Array.fill(town.length)(Array.fill(day + 1)(-1))

    def solve: Double = {
      town.map(_.count(_ == 1))
        .zipWithIndex
        .foreach { case (l, i) =>
          neighbors(i) = l
        }

      hide(village, day)
    }

    def hide(m: Int, d: Int): Double =
      if (d == 0) {
        if (m == prison) 1.0 else 0.0
      } else if (cache(m)(d) > -1) {
        cache(m)(d)
      } else {
        val probability: Double = town(m)
          .zipWithIndex
          .filter(_._1 == 1)
          .map { case (n, i) =>
            hide(i, d - 1) / neighbors(i).toDouble
          }
          .sum

        cache(m)(d) = probability
        cache(m)(d)
      }
  }

}
